versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java

```diff
@@ -0,0 +1,401 @@
+  }
+
+  @Override
+  protected void writeKeyListEntities(
+      NonTransactionalOperationContext ctx, List<KeyListEntity> newKeyListEntities) {
+    for (KeyListEntity keyList : newKeyListEntities) {
+      try {
+        insert(client.getKeyLists(), keyList.getId(), toProto(keyList.getKeys()).toByteArray());
+      } catch (ReferenceConflictException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+  }
+
+  @Override
+  protected void writeGlobalCommit(NonTransactionalOperationContext ctx, GlobalStateLogEntry entry)
+      throws ReferenceConflictException {
+    String id = toId(Hash.of(entry.getId()));
+    insert(client.getGlobalLog(), toDoc(id, entry.toByteArray()));
+  }
+
+  @Override
+  protected void unsafeWriteGlobalPointer(
+      NonTransactionalOperationContext ctx, GlobalStatePointer pointer) {
+    Document doc = toDoc(pointer);
+
+    UpdateResult result =
+        client
+            .getGlobalPointers()
+            .updateOne(
+                Filters.eq((Object) globalPointerKey),
+                new Document("$set", doc),
+                new UpdateOptions().upsert(true));
+
+    if (!result.wasAcknowledged()) {
+      throw new IllegalStateException(
+          "Unacknowledged write to " + client.getGlobalPointers().getNamespace());
+    }
+  }
+
+  @Override
+  protected boolean globalPointerCas(
+      NonTransactionalOperationContext ctx,
+      GlobalStatePointer expected,
+      GlobalStatePointer newPointer) {
+    Document doc = toDoc(newPointer);
+    byte[] expectedGlobalId = expected.getGlobalId().toByteArray();
+
+    UpdateResult result =
+        client
+            .getGlobalPointers()
+            .updateOne(
+                Filters.and(Filters.eq(globalPointerKey), Filters.eq("globalId", expectedGlobalId)),
+                new Document("$set", doc));
+
+    return result.wasAcknowledged()
+        && result.getMatchedCount() == 1
+        && result.getModifiedCount() == 1;
+  }
+
+  @Override
+  protected void cleanUpCommitCas(
+      NonTransactionalOperationContext ctx,
+      Hash globalId,
+      Set<Hash> branchCommits,
+      Set<Hash> newKeyLists) {
+    client.getGlobalLog().deleteOne(Filters.eq(toId(globalId)));
+
+    delete(client.getCommitLog(), branchCommits);
+    delete(client.getKeyLists(), newKeyLists);
+  }
+
+  @Override
+  protected GlobalStatePointer fetchGlobalPointer(NonTransactionalOperationContext ctx) {
+    return loadById(client.getGlobalPointers(), globalPointerKey, GlobalStatePointer::parseFrom);
+  }
+
+  @Override
+  protected GlobalStateLogEntry fetchFromGlobalLog(NonTransactionalOperationContext ctx, Hash id) {
+    return loadById(client.getGlobalLog(), id, GlobalStateLogEntry::parseFrom);
+  }
+
+  @Override
+  protected List<GlobalStateLogEntry> fetchPageFromGlobalLog(
+      NonTransactionalOperationContext ctx, List<Hash> hashes) {
+    return fetchPage(client.getGlobalLog(), hashes, GlobalStateLogEntry::parseFrom);
+  }
+
+  @FunctionalInterface
```

Comment:
Can you move this interface to `ProtoSerialization` so we can reuse it?

Reply:
Moved: dbf2425824c655c50cc7d19bc8df6fe742afff59